### PR TITLE
Add fast image processor for  ViViT 

### PR DESCRIPTION
**Pull request by msminhas93**
## What does this PR do?

This PR implements a fast PyTorch image processor for ViViT (`VivitImageProcessorFast`) that extends the existing `BaseImageProcessorFast` class. The new processor provides optimized image processing capabilities specifically designed for video understanding tasks using the ViViT model.

## Key Features

- **Fast PyTorch-native implementation**: Leverages PyTorch operations for efficient video frame processing
- **Video structure handling**: Supports nested video inputs (batch of videos, each video is a list of frames)
- **Custom rescaling with offset**: Implements ViViT's unique rescaling behavior with optional offset parameter for [-1, 1] range
- **Channel dimension handling**: Automatic detection and conversion between different channel formats (CHW vs HWC)
- **Optimized preprocessing pipeline**: Sequential application of resize, center crop, rescale, and normalize operations
- **Batch processing support**: Efficient handling of multiple video inputs with proper tensor stacking
- **AutoImageProcessor integration**: Properly registered in the auto mapping system for seamless usage

## Before submitting

- [x] This PR implements a new fast image processor for ViViT
- [x] Did you read the contributor guideline, Pull Request section?
- [x] Was this discussed/approved via a Github issue or the forum? Please add a link to it if that's the case. [Contributions Welcome] Add Fast Image Processors #36978
- [x] Did you make sure to update the documentation with your changes? Here are the documentation guidelines, and here are tips on formatting docstrings.
- [x] Did you write any new necessary tests? Updated test suite to run against both slow and fast processors with 23/23 tests passing

## Who can review?

@yonigozlan

### CODE REVIEW

**Comment by msminhas93** [comment]
Hi @yonigozlan! Could you please take a look. Thank you.

File: docs/source/en/model_doc/vivit.md [hunk 1]
```diff
@@ -9,7 +9,6 @@ Unless required by applicable law or agreed to in writing, software distributed
 an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 specific language governing permissions and limitations under the License.
 -->
-*This model was released on 2021-03-29 and added to Hugging Face Transformers on 2023-07-11.*
```
Comment (yonigozlan) [inline docs/source/en/model_doc/vivit.md]: Why remove?

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+def make_batched(videos) -> list[list]:
+    """Make videos batched for processing."""
+    if isinstance(videos, (list, tuple)) and isinstance(videos[0], (list, tuple)):
+        return videos
+    elif isinstance(videos, (list, tuple)):
+        return [videos]
+    else:
+        return [[videos]]
+
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: We have make_nested_list_of_images in image_utils for that

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        do_normalize (`bool`, *optional*, defaults to `True`):
+            Whether to normalize the image. Can be overridden by the `do_normalize` parameter in the `preprocess`
+            method.
+        image_mean (`float` or `list[float]`, *optional*, defaults to `IMAGENET_STANDARD_MEAN`):
+            Mean to use if normalizing the image. This is a float or list of floats the length of the number of
+            channels in the image. Can be overridden by the `image_mean` parameter in the `preprocess` method.
+        image_std (`float` or `list[float]`, *optional*, defaults to `IMAGENET_STANDARD_STD`):
+            Standard deviation to use if normalizing the image. This is a float or list of floats the length of the
+            number of channels in the image. Can be overridden by the `image_mean` parameter in the `preprocess` method.
+    """
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: No need, handled by auto_docstring
```suggestion
```

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        image = image.to(torch.float32)
+
+        # Use the most numerically stable approach
+        if offset:
+            # Compute (image * scale - 1) in one operation to minimize precision loss
+            rescaled_image = torch.sub(torch.mul(image, scale), 1.0)
+        else:
+            rescaled_image = torch.mul(image, scale)
+
+        return rescaled_image
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: This is way too complicated, we only have tensors at this point, so no need for all that. Also rescale_and_normalize from the parent should be modified to include offset as well

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                    frame = frame.unsqueeze(0)  # (H, W) -> (1, H, W)
+
+                if device is not None:
+                    frame = frame.to(device)
+
+                processed_frames.append(frame)
+
+            processed_videos.append(processed_frames)
+
+        return processed_videos
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: No need to override this. Just override _prepare_images_structure from the parent class, and call make_nested_list_of_images

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+                video_tensor = torch.stack(processed_frames, dim=0)
+                processed_videos.append(video_tensor)
+            else:
+                processed_videos.append(processed_frames)
+
+        # Stack videos if returning tensors
+        if return_tensors:
+            processed_videos = torch.stack(processed_videos, dim=0)
+
+        return BatchFeature(data={"pixel_values": processed_videos}, tensor_type=return_tensors)
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: You can process fully by batch still. just use the code from the parent class, and set `is_nested=True` in `group_images_by_shape` and `reorder_images`

File: src/transformers/models/vivit/image_processing_vivit_fast.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                    - `TensorType.TENSORFLOW` or `'tf'`: Return a batch of type `tf.Tensor`.
+                    - `TensorType.PYTORCH` or `'pt'`: Return a batch of type `torch.Tensor`.
+                    - `TensorType.NUMPY` or `'np'`: Return a batch of type `np.ndarray`.
+                    - `TensorType.JAX` or `'jax'`: Return a batch of type `jax.numpy.ndarray`.
+            data_format (`ChannelDimension` or `str`, *optional*, defaults to `ChannelDimension.FIRST`):
+                The channel dimension format for the output image. Can be one of:
+                    - `ChannelDimension.FIRST`: image in (num_channels, height, width) format.
+                    - `ChannelDimension.LAST`: image in (height, width, num_channels) format.
+                    - Unset: Use the inferred channel dimension format of the input image.
+        """
```
Comment (yonigozlan) [inline src/transformers/models/vivit/image_processing_vivit_fast.py]: No need to repeat that, also let's put this near the top of the class, just below the init
```suggestion
```

File: tests/models/vivit/test_image_processing_vivit.py [hunk 8]
```diff
@@ -106,123 +119,154 @@ def image_processor_dict(self):
...
+            encoded_videos = image_processing(video_inputs[0], return_tensors="pt").pixel_values
+            expected_output_video_shape = self.image_processor_tester.expected_output_image_shape([encoded_videos[0]])
+            self.assertEqual(tuple(encoded_videos.shape), (1, *expected_output_video_shape))
+
+            # Test batched
+            encoded_videos = image_processing(video_inputs, return_tensors="pt").pixel_values
+            expected_output_video_shape = self.image_processor_tester.expected_output_image_shape(encoded_videos)
+            self.assertEqual(
+                tuple(encoded_videos.shape), (self.image_processor_tester.batch_size, *expected_output_video_shape)
+            )
```
Comment (yonigozlan) [inline tests/models/vivit/test_image_processing_vivit.py]: Why was this added?

**Review by yonigozlan — COMMENTED** [review]
Hi @msminhas93, thanks for contributing, a few things can be simplified!

File: docs/source/en/model_doc/vivit.md (hunk 1 above)
Comment (msminhas93) [reply docs/source/en/model_doc/vivit.md]: I think I removed this accidentally. I've reverted the change

File: tests/models/vivit/test_image_processing_vivit.py (hunk 8 above)
Comment (msminhas93) [reply tests/models/vivit/test_image_processing_vivit.py]: I think I must have copy pasted this twice, removed it

**Comment by msminhas93** [comment]
Thank you for the comments! I've updated my PR @yonigozlan

**Comment by msminhas93** [comment]
@yonigozlan If you could please look at the PR that'll be awesome

Thank you!
